### PR TITLE
Ignore records from SQS which where seen in S3

### PR DESCRIPTION
**Pull request by AlexeyBarabash**
There is a situation when sync lib sends duplicating records, first from S3 and then from SQS, it may happens if syncing bookmarks early after establishing a sync chain:
```
[26515:26515:0426/170809.283612:INFO:CONSOLE(5175)] "listObjects (1) S3 content.timestamp=1556287688729", source: chrome-extension://nomlkjnggnifocmealianaaiobmebgil/extension/brave-sync/bundles/bundle.js (5175)
[26515:26515:0426/170809.286318:INFO:CONSOLE(5175)] "listObjects (1) S3 content.timestamp=1556287688737", source: chrome-extension://nomlkjnggnifocmealianaaiobmebgil/extension/brave-sync/bundles/bundle.js (5175)

[26515:26515:0426/170908.000107:INFO:CONSOLE(5116)] "SQS record timestamp=1556287706300", source: chrome-extension://nomlkjnggnifocmealianaaiobmebgil/extension/brave-sync/bundles/bundle.js (5116)
[26515:26515:0426/170908.288370:INFO:CONSOLE(5116)] "SQS record timestamp=1556287688729", source: chrome-extension://nomlkjnggnifocmealianaaiobmebgil/extension/brave-sync/bundles/bundle.js (5116)
[26515:26515:0426/170908.588513:INFO:CONSOLE(5116)] "SQS record timestamp=1556287688737", source: chrome-extension://nomlkjnggnifocmealianaaiobmebgil/extension/brave-sync/bundles/bundle.js (5116)
```
This breaks merging.
On brave-browser this was workarounded with https://github.com/brave/brave-core/pull/2016 (Don't send bookmarks) , but this seems not enough.

This PR does ignores records from SQS which were seen as from S3 by unique timestampt. If these records are processed twice in between "legal" records, this may cause mess.

STR for issue using brave-syncer branch:
1. device a create `folder A` and `bookmark sync`
2. device b connect to it and move sync under A
3. device a resolves records to 0, no changes reflected

### CODE REVIEW

File: lib/s3Helper.js [hunk 1]
```diff
@@ -166,7 +173,10 @@ function listNotificationsRecursively (SQS, options, category, prefix, currentCo
           let contentMessage = {
             Key: key
           }
-          currentContent.push(contentMessage)
+          var timestapmt = extractTimeStampt(contentMessage);
```
Comment (SergeyZhukovsky) [inline lib/s3Helper.js]: typo `timestampt`

**Comment by SergeyZhukovsky** [comment]
Could it be that 2 records have exactly the same `timestamp`? Maybe we should replace it to use an `objectId` instead?

**Comment by diracdeltas** [comment]
it would be good to have test coverage for this. https://github.com/brave/sync/blob/staging/test/s3Helper.js

File: lib/s3Helper.js (hunk 1 above)
Comment (AlexeyBarabash) [reply lib/s3Helper.js]: thanks, fixed this

**Comment by AlexeyBarabash** [comment]
> Could it be that 2 records have exactly the same `timestamp`? Maybe we should replace it to use an `objectId` instead?

@SergeyZhukovsky timestamp is the most close field by purpose to unique identify the record. I hadn't seen the case when timestampts are exactly the same, but in theory that can rare happen.

objectId is not suitable for this because it identifies an object and not the operation.

I take timestampt from Key, which looks like
Key=`0/XVlVvDI3dlb7Dvk6CkMWRN2piY3I7CJrlPUta7seG/s=/0/1556309475075/0/NF5Ra/Cs8CNRKja74XBGnqcLSog1yVoCTwGbg6sxhqzb15jt1pEVeF7flavfdBqqBFn38AMtqyxD/xQks7yikLA2DLc7VwVe8z+wK8sJ9q0CCCIgO2SwV1kGfMFMH8lS22HX6Kc5QCnj5Js3S1BKcdgeX+uIa8GEUCGHErbQEDRqbqoy596j7uboLXMmCC/Yo80EurmqX5Ga5CxRM4PlIAUSUj39s3k3oy8YBWcWBb+hRF118kuao0b5sKnOg+fkIf68X5AZIIhozp7DHG1zoMypFy9YJZI3YSwmA/xp2cE0YmWFWOiAuJD+kOzFA4UyMVtEsnFpEyu8RhMpzowmCCDq8S0kztI+iq/IunpIF5itOkAlbXxeNREHgUWOgAlqTtRZWKmIwh8SwliJJwc+X2iFEFqGVm9FQP6/xAwIV8umOe0yc9c7E9DIwU+sTPySaVg54QCCYaGAAAExmUma+2EiKKNs6gvhO7GBz94JYAAA==`
And this is exactly what is unique, but it is too big to use it and store without erase policy.

**Comment by AlexeyBarabash** [comment]
> it would be good to have test coverage for this. https://github.com/brave/sync/blob/staging/test/s3Helper.js

@diracdeltas thanks, looking how to make a testcase

**Comment by AlexeyBarabash** [comment]
closed this pr because there is a better way

**Comment by AlexeyBarabash** [comment]
I had removed commit 
https://github.com/brave/sync/pull/306/commits/a8fa6a4a4c9dc38aabef7ac40cabbe84d22d4d10
`Ignore messages from ourself` because clients use this info to ensure the record was successfully arrived to Amazon storage. 
But this should be filtered out on the client, client should not answer `resolve-sync-records` when record has a device_id equal to the device_id of the client.

**Comment by AlexeyBarabash** [comment]
@SergeyZhukovsky @diracdeltas @darkdh 
Can I get another review?

Fixed according to the previous notices:
I use timeStamp+crc as a key;
a testcase was added.

File: lib/s3Helper.js [hunk 2]
```diff
@@ -93,6 +93,15 @@ function compareS3Key (a, b) {
   }
 }
 
+module.exports.extractPseudoKey = function (a) {
+  const otherParts = a.Key.substring(47).split('/')
```
Comment (darkdh) [inline lib/s3Helper.js]: can we use `parseS3Key` here instead of duplicating parsing logic?

File: test/s3Helper.js [hunk 3]
```diff
@@ -16,6 +16,16 @@ test('s3Helper', (t) => {
...
   })
 
+  t.test('extractPseudoKey', (t) => {
+    t.plan(2)
+    const s3Prefix = '0/xGrUe8vokl9kjAx+RTu9t6I1UnOT7mcdcizAVI+2bos=/2/1482435340000'
+    const data = new Uint8Array(10)
+    const s3Keys = s3Helper.encodeDataToS3KeyArray(s3Prefix, data)
+    t.equal(s3Keys.length, 1, `${t.name} encodes to one part`)
+    const pseudoKey = s3Helper.extractPseudoKey({ Key: s3Keys[0] })
+    t.equal(pseudoKey, '1482435340000-3ZYcXs', `${t.name} expected pseudo-key`)
```
Comment (darkdh) [inline test/s3Helper.js]: is it possible to get `3ZYcXs` by calculating it like https://github.com/brave/sync/pull/306/files#diff-33e6fea82e3ab87ec881daae3c0c6384R41 ?

File: lib/s3Helper.js (hunk 2 above)
Comment (AlexeyBarabash) [reply lib/s3Helper.js]: Yes, sure. Fixed this.

File: test/s3Helper.js (hunk 3 above)
Comment (AlexeyBarabash) [reply test/s3Helper.js]: Thanks for pointing. Fixed.

**Pull request merged**